src/validation_layer.h

```diff
@@ -0,0 +1,144 @@
+    const bool block_valid;
+
+    BlockValidationResponse(bool _block_valid, bool _is_new)
+        : is_new(_is_new), block_valid(_block_valid){};
+};
+
+/**
+ * Encapsulates a request to validate a block
+ */
+class BlockValidationRequest : public ValidationRequest<BlockValidationResponse>
+{
+    friend ValidationLayer;
+
+private:
+    BlockValidationRequest(ValidationLayer& validation_layer, const std::shared_ptr<const CBlock> block, bool force_processing, const std::function<void()> on_ready)
+        : m_validation_layer(validation_layer), m_block(block), m_force_processing(force_processing), m_on_ready(on_ready){};
+
+    //! Does the validation
+    void operator()() override;
+
+    //! Returns a block hash
+    std::string GetId() const override;
+
+    const ValidationLayer& m_validation_layer;
+
+    //! The block to be validated
+    const std::shared_ptr<const CBlock> m_block;
+
+    //! Was this block explicitly requested (currently required by ProcessNewBlock)
+    const bool m_force_processing;
+
+    //! A callback to invoke when ready
+    //! This is a workaround because c++11 does not support multiplexed waiting on futures
+    //! In a move to subsequent standards when this behavior is supported this can probably be removed
+    const std::function<void()> m_on_ready;
+};
+
+/**
+ * Public interface to block validation
+ *
+ * Two apis:
+ *  - asynchronous: SubmitForValidation(object) -> future<Response>
+ *  - synchronous:  Validate(object) -> Response (just calls SubmitForValidation and blocks on the response)
+ *
+ * Internally, a validation thread pulls validations requests from a queue, processes them and satisfies the promise
+ * with the result of validation.
+ */
+class ValidationLayer
+{
+    friend BlockValidationRequest;
+
+    typedef WorkQueue<WorkerMode::BLOCKING> ValidationQueue;
+    typedef ConsumerThread<WorkerMode::BLOCKING> ValidationThread;
+
+public:
+    ValidationLayer(const CChainParams& chainparams)
+        : m_chainparams(chainparams), m_validation_queue(std::make_shared<ValidationQueue>(100)) {}
+    ~ValidationLayer(){};
+
+    //! Starts the validation layer (creating the validation thread)
+    void Start();
+
+    //! Stops the validation layer (stopping the validation thread)
+    void Stop();
+
+    //! Submit a block for asynchronous validation
+    std::future<BlockValidationResponse> SubmitForValidation(const std::shared_ptr<const CBlock> block, bool force_processing, std::function<void()> on_ready = []() {});
+
+    //! Submit a block for validation and block on the response
+    BlockValidationResponse Validate(const std::shared_ptr<const CBlock> block, bool force_processing);
+
+private:
+    //! Internal utility method - sets up and calls ProcessNewBlock
+    BlockValidationResponse ValidateInternal(const std::shared_ptr<const CBlock> block, bool force_processing) const;
```

Comment:
Any reason this doesn't just live on BlockValidationRequest? Could avoid some friendliness that way.